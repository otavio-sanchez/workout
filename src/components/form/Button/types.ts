export interface Props {
    onClick?: any;
    text?: string;
    active?: boolean;
}
