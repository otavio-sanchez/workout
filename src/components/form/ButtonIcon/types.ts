export interface Props {
    onClick?: any;
    children: JSX.Element;
    active?: boolean;
}
