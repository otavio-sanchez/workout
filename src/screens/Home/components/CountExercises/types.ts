import { Form } from '../FormWorkout/types'

export interface Props {
    list: Form[]
}