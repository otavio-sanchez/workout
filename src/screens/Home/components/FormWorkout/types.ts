export interface Props {
    onSubmit: Function;
    title: string;
}

export interface Form {
    time: string
    date: string,
    activities: string,
}