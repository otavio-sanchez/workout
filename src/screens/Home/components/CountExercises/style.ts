
import styled from 'styled-components';

const Content = styled.div`
    position: flex;
    width: 100%;
    padding: 18px;
    box-sizing: border-box;
    justify-content: center;
    align-items: center;
    text-align: center;
`;


export { Content }
