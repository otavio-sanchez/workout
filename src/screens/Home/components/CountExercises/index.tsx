import React, { useEffect, useState, useCallback } from "react";
import { Col, Container, Row } from "../../../../components/grid";
import { Props } from "./types";
import { Form } from "../FormWorkout/types";
import { Text } from "../../../../components/typography";
import { Content } from './style'
 
const CountExercises = ({ list }: Props): JSX.Element => {
  const [hours, setHours] = useState(0);

  const countHours = (list: Form[]) => {
    const total = list.reduce((a, b): number => a + parseFloat(b.time), 0);

    setHours(total);
  };

  useEffect(() => {
    countHours(list);
  }, [list]);

  return (
    <Container>
      <Row columnsDesktop={1} columnsTablet={1} columnsMobile={1}>
        <Col>
          <Content>
            <Text type="regular">
              {hours > 0 &&
                `${hours.toFixed(1).toString()} ${
                  hours > 1 ? "hours" : "hour"
                } of exercise`}
            </Text>
          </Content>
        </Col>
      </Row>
    </Container>
  );
};

export default CountExercises;
