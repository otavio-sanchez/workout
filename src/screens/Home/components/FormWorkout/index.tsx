import React, { useState, useEffect, FormEvent } from "react";
import {
  Form,
  TextInput,
  Button,
  Datepicker,
  Select,
} from "../../../../components/form";
import { Col, Container, Row } from "../../../../components/grid";
import { Props, Form as FormType } from "./types";
import { validation, formErrors } from "./validation";
import { exercises as mockExercises } from "../../../../mock/exercises";
import { ToastContainer, toast } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";

const FormWorkout = ({ title, onSubmit }: Props): JSX.Element => {
  const initState: FormType = {
    time: "",
    date: new Date().toString(),
    exercises: "",
  };

  const initStateErrors: any = {};

  const [form, setForm] = useState(initState);
  const [errors, setErrors] = useState(initStateErrors);

  const onChange = (name: string, value: string) => {
    setForm({ ...form, [`${name}`]: value });
  };

  const submit = async () => {
    const validate = await validation(form);

    setErrors(initStateErrors);
    if (validate) {
      toast('Exercise added')
      onSubmit(form);
      setForm(initState);
    } else {
      const formError = await formErrors(form);

      if (formError) {
        setErrors(formError);
      }
    }
  };

  return (
    <>
      <Form title={title}>
        <Container>
          <Row columnsDesktop={4} columnsTablet={1} columnsMobile={1}>
            <Col>
              <TextInput
                name="time"
                type="number"
                value={form["time"]}
                onChange={onChange}
                error={errors && errors["time"]}
                placeholder={"Hours"}
                
              />
            </Col>
            <Col>
              <Datepicker onChange={onChange} name="date" />
            </Col>
            <Col>
              <Select
                name="exercises"
                onChange={onChange}
                label={"Select Exercise"}
                value={form["exercises"]}
                options={mockExercises}
                error={errors && errors["exercises"]}
              />
            </Col>
            <Col>
              <Button type="submit" onClick={() => submit()} text={"Add"} />
            </Col>
          </Row>
        </Container>
      </Form>
      <ToastContainer />
    </>
  );
};

export default FormWorkout;
