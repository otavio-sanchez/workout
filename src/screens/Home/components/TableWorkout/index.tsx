import React, { useState, useEffect } from "react";

import {
  Table,
  TableRow,
  TableCol,
  TableColHead,
} from "../../../../components/table";

import { ButtonIcon } from "../../../../components/form";
import Icon from "../../../../assets/theme/icons/subtrair.svg";


import { Props } from "./types";

import { formatDate } from "../../../../utils/replace";
import { Block, ArrowIcon } from "./styles";

const Home = ({ list, removeExercise }: Props): JSX.Element => {
  const [listExercises, setListExercises] = useState([]);

  useEffect(() => {
    orderDate()
    setListExercises(list);
  }, [list]);

  const [itemActive, setItemActive] = useState({
    number: null,
    active: false,
  });

  const [dateAsc, setDateAsc] = useState(true);
  const [timeAsc, setTimeAsc] = useState(true);
  const [typeAsc, setTypeAsc] = useState(true);

  const orderTime = () => {
    setTimeAsc(!timeAsc);
    setListExercises(
      listExercises.sort((a, b) => {
        if (a.time > b.time) {
          return timeAsc ? 1 : -1;
        }
        if (a.time < b.time) {
          return timeAsc ? -1 : 1;
        }

        return 0;
      })
    );
  };

  const orderDate = () => {
    setDateAsc(!dateAsc);
    setListExercises(
      listExercises.sort((a, b) => {
        if (new Date(a.date) > new Date(b.date)) {
          return dateAsc ? 1 : -1;
        }
        if (new Date(a.date) < new Date(b.date)) {
          return dateAsc ? -1 : 1;
        }

        return 0;
      })
    );
  };

  const orderType = () => {
    setTypeAsc(!typeAsc);
    setListExercises(
      listExercises.sort((a, b) => {
        if (a.type > b.type) {
          return typeAsc ? 1 : -1;
        }
        if (a.type < b.type) {
          return typeAsc ? -1 : 1;
        }

        return 0;
      })
    );
  };

  return list.length ? (
    <Table
      header={
        <>
          <TableColHead>
            <>
              Time
              <ButtonIcon onClick={orderTime}>
                <ArrowIcon active={timeAsc} />
              </ButtonIcon>
            </>
          </TableColHead>
          <TableColHead>
            <>
              Type
              <ButtonIcon onClick={orderType}>
                <ArrowIcon  active={typeAsc} />
              </ButtonIcon>
            </>
          </TableColHead>
          <TableColHead>
            <>
              Date
              <ButtonIcon onClick={() => orderDate()}>
                <ArrowIcon  active={dateAsc} />
              </ButtonIcon>
            </>
          </TableColHead>
          <TableColHead> </TableColHead>
        </>
      }
    >
      {listExercises.map((item, index) => (
        <TableRow
          key={index}
          active={(active: boolean) =>
            setItemActive({
              number: index,
              active,
            })
          }
        >
          <TableCol>{item.time}</TableCol>
          <TableCol>{item.activities}</TableCol>
          <TableCol>{formatDate(item.date)}</TableCol>
          <TableCol>
            {itemActive.number === index ? (
              <ButtonIcon
                onClick={() => {
                  setItemActive({
                    number: null,
                    active: false,
                  });
                  removeExercise(item);
                }}
              >
                <Icon />
              </ButtonIcon>
            ) : (
              <Block />
            )}
          </TableCol>
        </TableRow>
      ))}
    </Table>
  ) : (
    <>Empty</>
  );
};

export default Home;
