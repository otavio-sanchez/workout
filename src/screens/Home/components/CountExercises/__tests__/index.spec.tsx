import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import CountExercises from "../../CountExercises";
import { Props } from "../types";
import { listExercises } from "../../../../../constants/mock/exercises";
import { Text } from "../../../../../components/typography";
import { withHooks } from "jest-react-hooks-shallow";

describe("CountExercises", () => {
  const props: Props = {
    list: listExercises,
  };


  it("should render CountExercises component", () => {
    const wrapper = shallow(<CountExercises {...props} />);
    expect(wrapper.find(Text)).toHaveLength(1);
  });

  it("should render feedback when there count data", () => {
    withHooks(() => {
      const wrapper = shallow(<CountExercises {...props} />);
      wrapper.setProps({ list: listExercises });
      expect(wrapper.find(Text).get(0).props.children).toContain(
        "3.0 hours of exercise"
      );
    });
  });

  it("should render feedback when there is not data", () => {
    const wrapper = shallow(<CountExercises {...props} />);
    wrapper.setProps({ list: [] });
    expect(wrapper.find(Text).get(0).props.children).toContain(" ");
  });

  it("should render with return count", () => {
    const wrapper = mount(<CountExercises {...props} />);
    expect(wrapper.find(Text).length).toEqual(1);
  });
});
