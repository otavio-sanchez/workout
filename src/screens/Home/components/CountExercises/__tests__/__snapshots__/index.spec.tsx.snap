// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountExercises renders without crashing given the required props 1`] = `
<Container>
  <Row
    columnsDesktop={1}
    columnsMobile={1}
    columnsTablet={1}
  >
    <Col>
      <styled.div>
        <Text
          type="regular"
        >
          3.0 hours of exercise
        </Text>
      </styled.div>
    </Col>
  </Row>
</Container>
`;
